### PR TITLE
rounded rectangle graphics

### PR DESCRIPTION
**Pull request by gintsgints**
This is my attempt to add functionality for additional canvas shape, rounded rectangle.

### CODE REVIEW

**Comment by vladh** [comment]
I don't think this approach is flexible enough. For example, I would expect drawing a square with a very high border radius to result in a circle. This is what happens in [`iced::Border::rounded`](https://docs.iced.rs/iced/struct.Border.html#method.rounded) and also in CSS. However, in your implementation, a buggy shape is drawn.

**Comment by gintsgints** [comment]
Hmm, so what should flexible solution do instead? 
I can make radius relative to the width of the rectangle or disallow too big radius.

**Comment by vladh** [comment]
> Hmm, so what should flexible solution do instead?

It looks like, at the very least, you want something like:

```rust
radius = min(radius, min(height / 2, width / 2))
``` 

See:
* https://www.w3.org/TR/css-backgrounds-3/#the-border-radius
* https://www.w3.org/Style/CSS/Tracker/issues/29

It would also be nice to be able to specify horizontal and vertical radius, and to specify individual radii for each of the 4 corners, by using or extending the existing [`iced::Border::Radius`](https://docs.iced.rs/iced/border/struct.Radius.html).

**Comment by gintsgints** [comment]
One commit implement just min check, other use Radius as argument, so it is possible to specify different radius for each corner.

File: core/src/border.rs [hunk 1]
```diff
@@ -54,7 +54,7 @@ impl Border {
 /// The border radii for the corners of a graphics primitive in the order:
 /// top-left, top-right, bottom-right, bottom-left.
 #[derive(Debug, Clone, Copy, PartialEq, Default)]
-pub struct Radius([f32; 4]);
+pub struct Radius(pub [f32; 4]);
```
Comment (vladh) [inline core/src/border.rs:57]: I don't think this needs to be `pub` — the `from()` methods can be used.

**Review by vladh — CHANGES_REQUESTED** [review]

File: core/src/border.rs (hunk 1 above)
Comment (gintsgints) [reply core/src/border.rs]: Sorry, you are right. Fixed.

File: graphics/src/geometry/path.rs [hunk 2]
```diff
@@ -47,6 +47,12 @@ impl Path {
         Self::new(|p| p.rectangle(top_left, size))
     }
 
+    /// Creates a new [`Path`] representing a rounded rectangle given its top-left
+    /// corner coordinate, its `Size` and radius.
```
Comment (vladh) [inline graphics/src/geometry/path.rs:51]: Maybe:

```
its [`Size`] and border [`Radius`]
```

File: graphics/src/geometry/path/builder.rs [hunk 3]
```diff
@@ -160,6 +161,51 @@ impl Builder {
         self.close();
     }
 
+    /// Adds a rounded rectangle to the [`Path`] given its top-left
+    /// corner coordinate its `Size` and curve radius.
```
Comment (vladh) [inline graphics/src/geometry/path/builder.rs:165]: Maybe:

```
its [`Size`] and border [`Radius`]
```

**Review by vladh — CHANGES_REQUESTED** [review]

**Comment by vladh** [comment]
This looks okay to me right now, but I'm not a maintainer so you'll have to wait for one to look at this.

File: graphics/src/geometry/path.rs (hunk 2 above)
Comment (gintsgints) [reply graphics/src/geometry/path.rs]: Fixed

File: graphics/src/geometry/path/builder.rs (hunk 3 above)
Comment (gintsgints) [reply graphics/src/geometry/path/builder.rs]: fixed

**Review by hecrj — APPROVED** [review]
Thanks!

I have removed the example since we already have too many.

**Pull request merged**